vmm/src/console_devices.rs

```diff
@@ -0,0 +1,300 @@
+            // they will be closed on a reboot and the numbers reused
+
+            // SAFETY: FFI call to dup. Trivially safe.
+            let stdout = unsafe { libc::dup(libc::STDOUT_FILENO) };
+            if stdout == -1 {
+                return vmm_sys_util::errno::errno_result().map_err(ConsoleDeviceError::DupFd);
+            }
+            // SAFETY: stdout is valid and owned solely by us.
+            let stdout = unsafe { File::from_raw_fd(stdout) };
+
+            // SAFETY: FFI call. Trivially safe.
+            if unsafe { libc::isatty(libc::STDOUT_FILENO) } == 1 {
+                vmm.console_resize_pipe = Some(
+                    listen_for_sigwinch_on_tty(
+                        stdout.try_clone().unwrap(),
+                        &vmm.seccomp_action,
+                        vmm.hypervisor.hypervisor_type(),
+                    )
+                    .map_err(ConsoleDeviceError::StartSigwinchListener)?,
+                );
+            }
+
+            // Make sure stdout is in raw mode, if it's a terminal.
+            set_raw_mode(&stdout, vmm.original_termios_opt.clone())?;
+            console_info.console_main_fd = Some(stdout.into_raw_fd());
+        }
+        ConsoleOutputMode::Null | ConsoleOutputMode::Socket | ConsoleOutputMode::Off => {}
+    }
+
+    match vmconfig.serial.mode {
+        ConsoleOutputMode::File => {
+            let file = File::create(vmconfig.serial.file.as_ref().unwrap())
+                .map_err(ConsoleDeviceError::CreateConsoleDevice)?;
+            console_info.serial_main_fd = Some(file.into_raw_fd());
+        }
+        ConsoleOutputMode::Pty => {
+            let (main_fd, sub_fd, path) =
+                create_pty().map_err(ConsoleDeviceError::CreateConsoleDevice)?;
+            console_info.serial_main_fd = Some(main_fd.into_raw_fd());
+            set_raw_mode(&sub_fd.as_raw_fd(), vmm.original_termios_opt.clone())?;
+            vmconfig.serial.file = Some(path.clone());
+        }
+        ConsoleOutputMode::Tty => {
+            // During vm_shutdown, when serial device is closed, FD#2(STDOUT)
+            // will be closed and FD#2 could be reused in a future boot of the
+            // guest by a different file.
+            // To ensure FD#2 always points to STANDARD OUT, a `dup` of STDOUT
+            // is passed to serial device. Doing so, even if the serial device
+            // were to be closed, FD#2 will continue to point to STANDARD OUT.
+
+            // SAFETY: FFI call to dup. Trivially safe.
+            let stdout = unsafe { libc::dup(libc::STDOUT_FILENO) };
+            if stdout == -1 {
+                return vmm_sys_util::errno::errno_result().map_err(ConsoleDeviceError::DupFd);
+            }
+            // SAFETY: stdout is valid and owned solely by us.
+            let stdout = unsafe { File::from_raw_fd(stdout) };
+
+            // SAFETY: FFI call. Trivially safe.
+            if unsafe { libc::isatty(libc::STDOUT_FILENO) } == 1 {
```

Comment:
This should check `stdout.as_raw_fd()`.

Reply:
`stdout` at this point does not have STDOUT_FILENO inside. Rather, it has a dup'ed fd from STDOUT_FILENO. Since you change stdout in the following block, you should check the correct fd, even though the end result is the same.